// @ngxs/store/esm5/ngxs-store.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgxsModule, Action, Store, State, Select, Actions, getSelectorMetadata, getStoreMetadata, ensureStoreMetadata, ensureSelectorMetadata, ofAction, ofActionDispatched, ofActionSuccessful, ofActionCanceled, ofActionErrored, Selector, getActionTypeFromInstance, actionMatcher, createSelector, NGXS_PLUGINS, StateStream, setValue, getValue, InitState, UpdateState } from './index';
export { InternalActions as ɵg, OrderedSubject as ɵf } from './src/actions-stream';
export { SelectFactory as ɵe } from './src/decorators/select';
export { InternalDispatchedActionResults as ɵn, InternalDispatcher as ɵo } from './src/internal/dispatcher';
export { StateContextFactory as ɵp } from './src/internal/state-context-factory';
export { StateFactory as ɵm } from './src/internal/state-factory';
export { InternalStateOperations as ɵq } from './src/internal/state-operations';
export { NgxsFeatureModule as ɵb, NgxsRootModule as ɵa, ROOT_OPTIONS as ɵd, ngxsConfigFactory as ɵc } from './src/module';
export { PluginManager as ɵr } from './src/plugin-manager';
export { FEATURE_STATE_TOKEN as ɵi, META_KEY as ɵj, NgxsConfig as ɵl, ROOT_STATE_TOKEN as ɵh, SELECTOR_META_KEY as ɵk } from './src/symbols';
