// @ngxs/store/esm2015/src/internal/state-context-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { setValue, getValue } from '../utils/utils';
import { InternalStateOperations } from '../internal/state-operations';
/**
 * State Context factory class
 * @ignore
 */
export class StateContextFactory {
    /**
     * @param {?} _internalStateOperations
     */
    constructor(_internalStateOperations) {
        this._internalStateOperations = _internalStateOperations;
    }
    /**
     * Create the state context
     * @param {?} metadata
     * @return {?}
     */
    createStateContext(metadata) {
        /** @type {?} */
        const root = this._internalStateOperations.getRootStateOperations();
        return {
            /**
             * @return {?}
             */
            getState() {
                /** @type {?} */
                const state = root.getState();
                return getValue(state, metadata.depth);
            },
            /**
             * @param {?} val
             * @return {?}
             */
            patchState(val) {
                /** @type {?} */
                const isArray = Array.isArray(val);
                /** @type {?} */
                const isPrimitive = typeof val !== 'object';
                if (isArray) {
                    throw new Error('Patching arrays is not supported.');
                }
                else if (isPrimitive) {
                    throw new Error('Patching primitives is not supported.');
                }
                /** @type {?} */
                const state = root.getState();
                /** @type {?} */
                const local = getValue(state, metadata.depth);
                /** @type {?} */
                const clone = Object.assign({}, local);
                for (const k in val) {
                    clone[k] = val[k];
                }
                /** @type {?} */
                const newState = setValue(state, metadata.depth, clone);
                root.setState(newState);
                return newState;
            },
            /**
             * @param {?} val
             * @return {?}
             */
            setState(val) {
                /** @type {?} */
                let state = root.getState();
                state = setValue(state, metadata.depth, val);
                root.setState(state);
                return state;
            },
            /**
             * @param {?} actions
             * @return {?}
             */
            dispatch(actions) {
                return root.dispatch(actions);
            }
        };
    }
}
StateContextFactory.decorators = [
    { type: Injectable }
];
/** @nocollapse */
StateContextFactory.ctorParameters = () => [
    { type: InternalStateOperations }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    StateContextFactory.prototype._internalStateOperations;
}
