// @ngxs/router-plugin/esm2015/src/router.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgxsModule } from '@ngxs/store';
import { DefaultRouterStateSerializer, RouterStateSerializer } from './serializer';
import { RouterState } from './router.state';
// NOTE: Must mark as `dynamic` due to
// https://github.com/dherges/ng-packagr/issues/767
/** @type {?} */
export const NgxsModuleRouterState = NgxsModule.forFeature([RouterState]);
// @dynamic
export class NgxsRouterPluginModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NgxsRouterPluginModule,
            providers: []
        };
    }
}
NgxsRouterPluginModule.decorators = [
    { type: NgModule, args: [{
                imports: [NgxsModuleRouterState],
                providers: [{ provide: RouterStateSerializer, useClass: DefaultRouterStateSerializer }]
            },] }
];
